cpp/src/parquet/encoding-test.cc

```diff
@@ -314,6 +325,258 @@ TEST(TestDictionaryEncoding, CannotDictDecodeBoolean) {
+  std::shared_ptr<::arrow::Array> expected_dense_;
+  int num_values_;
+  int null_count_;
+  vector<ByteArray> input_data_;
+  const uint8_t* valid_bits_;
+  std::unique_ptr<ByteArrayEncoder> encoder_;
+  std::unique_ptr<ByteArrayDecoder> decoder_;
+  std::shared_ptr<Buffer> buffer_;
+};
+
+// ----------------------------------------------------------------------
+// Arrow builder decode tests for PlainByteArrayDecoder
+class FromPlainEncoding : public TestDecodeArrow {
+ public:
+  void SetupEncoderDecoder() override {
+    encoder_ = MakeTypedEncoder<ByteArrayType>(Encoding::PLAIN);
+    decoder_ = MakeTypedDecoder<ByteArrayType>(Encoding::PLAIN);
+    ASSERT_NO_THROW(encoder_->Put(input_data_.data(), num_values_));
+    buffer_ = encoder_->FlushValues();
+    decoder_->SetData(num_values_, buffer_->data(), static_cast<int>(buffer_->size()));
+  }
+
+  void TearDown() override {}
+};
+
+TEST_P(FromPlainEncoding, DecodeDense) {
+  ::arrow::internal::ChunkedBinaryBuilder builder(static_cast<int>(buffer_->size()),
+                                                  default_memory_pool());
+  auto actual_num_values =
+      decoder_->DecodeArrow(num_values_, null_count_, valid_bits_, 0, &builder);
+  CheckDecodeDense(actual_num_values, builder);
+}
+
+TEST_P(FromPlainEncoding, DecodeNonNullDense) {
+  // Skip this test if input data contains nulls (optionals)
+  SKIP_TEST_IF(null_count_ > 0)
+  ::arrow::internal::ChunkedBinaryBuilder builder(static_cast<int>(buffer_->size()),
+                                                  default_memory_pool());
+  auto actual_num_values = decoder_->DecodeArrowNonNull(num_values_, &builder);
+  CheckDecodeDense(actual_num_values, builder);
+}
+
+TEST_P(FromPlainEncoding, DecodeBinaryDict) {
+  ::arrow::BinaryDictionaryBuilder builder(default_memory_pool());
+  auto actual_num_values =
+      decoder_->DecodeArrow(num_values_, null_count_, valid_bits_, 0, &builder);
+  CheckDecodeDict(actual_num_values, builder);
+}
+
+TEST_P(FromPlainEncoding, DecodeStringDict) {
+  ::arrow::StringDictionaryBuilder builder(default_memory_pool());
+  auto actual_num_values =
+      decoder_->DecodeArrow(num_values_, null_count_, valid_bits_, 0, &builder);
+  CheckDecodeDict(actual_num_values, builder);
+}
+
+TEST_P(FromPlainEncoding, DecodeNonNullBinaryDict) {
+  // Skip this test if input data contains nulls (optionals)
+  SKIP_TEST_IF(null_count_ > 0)
+  ::arrow::BinaryDictionaryBuilder builder(default_memory_pool());
+  auto actual_num_values = decoder_->DecodeArrowNonNull(num_values_, &builder);
+  CheckDecodeDict(actual_num_values, builder);
+}
+
+TEST_P(FromPlainEncoding, DecodeNonNullStringDict) {
+  // Skip this test if input data contains nulls (optionals)
+  SKIP_TEST_IF(null_count_ > 0)
+  ::arrow::StringDictionaryBuilder builder(default_memory_pool());
+  auto actual_num_values = decoder_->DecodeArrowNonNull(num_values_, &builder);
+  CheckDecodeDict(actual_num_values, builder);
+}
```

Comment:
Seems like the code duplication in these 4 test functions ought to be addressable